Add ssh-manager

7ym0n opened the pull request:
### Brief summary of what the package does

Use SSH to manage remote servers like xshell, mobaxterm or other tools.

### Direct link to the package repository

https://github.com/7ym0n/dotfairy-ssh-manager

### Your association with the package

I'm maintainer.

### Relevant communications with the upstream package maintainer

None needed

### Checklist

<!-- Please confirm with `x`: -->

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [ ] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

<!-- After submitting, please fix any problems the CI reports. -->

riscy commented:
Thanks for this -- please take a look at the checklist items:

### ssh-manager.el ###

byte-compile (using Emacs 27.1):
```
ssh-manager.el:217:11:Warning: `term-setup-hook' is an obsolete variable (as
    of 24.4); use either `emacs-startup-hook' or `tty-setup-hook' instead.
```

checkdoc (using version 0.6.2):
```
Warning (emacs): 
ssh-manager.el:48: All variables and subroutines might as well have a documentation string
Warning (emacs): 
ssh-manager.el:66: Argument `buf-name' should appear (as BUF-NAME) in the doc string
Warning (emacs): 
ssh-manager.el:72: Argument `session' should appear (as SESSION) in the doc string
Warning (emacs): 
ssh-manager.el:78: All variables and subroutines might as well have a documentation string
Warning (emacs): 
ssh-manager.el:84: First sentence should end with punctuation
Warning (emacs): 
ssh-manager.el:84: First line should be capitalized
Warning (emacs): 
ssh-manager.el:112: First sentence should end with punctuation
Warning (emacs): 
ssh-manager.el:119: First sentence should end with punctuation
Warning (emacs): 
ssh-manager.el:124: All interactive functions should have documentation
Warning (emacs): 
ssh-manager.el:149: First line is not a complete sentence
Warning (emacs): 
ssh-manager.el:149: Some lines are over 80 columns wide
Warning (emacs): 
ssh-manager.el:149: Argument `format' should appear (as FORMAT) in the doc string
Warning (emacs): 
ssh-manager.el:151: Name emacs should appear capitalized as Emacs
Warning (emacs): 
ssh-manager.el:155: Name emacs should appear capitalized as Emacs
Warning (emacs): 
ssh-manager.el:151: Name emacs should appear capitalized as Emacs
Warning (emacs): 
ssh-manager.el:215: First sentence should end with punctuation
Warning (emacs): 
ssh-manager.el:215: Argument `term-name' should appear (as TERM-NAME) in the doc string
Warning (emacs): 
ssh-manager.el:247: All variables and subroutines might as well have a documentation string
Warning (emacs): 
ssh-manager.el:324: First line should be capitalized
Warning (emacs): 
ssh-manager.el:332: Argument `session' should appear (as SESSION) in the doc string
Warning (emacs): 
ssh-manager.el:342: Argument `kind' should appear (as KIND) in the doc string
Warning (emacs): 
ssh-manager.el:393: Argument `kind' should appear (as KIND) in the doc string
Warning (emacs): 
ssh-manager.el:411: Argument `session' should appear (as SESSION) in the doc string
Warning (emacs): 
ssh-manager.el:426: Argument `session' should appear (as SESSION) in the doc string
Warning (emacs): 
ssh-manager.el:438: Argument `history' should appear (as HISTORY) in the doc string
Warning (emacs): 
ssh-manager.el:447: First sentence should end with punctuation
Warning (emacs): 
ssh-manager.el:469: First sentence should end with punctuation
Warning (emacs): 
ssh-manager.el:469: Argument `server' should appear (as SERVER) in the doc string
Warning (emacs): 
ssh-manager.el:528: All variables and subroutines might as well have a documentation string
Warning (emacs): 
ssh-manager.el:532: Some lines are over 80 columns wide
Warning (emacs): 
ssh-manager.el:559: Argument `server' should appear (as SERVER) in the doc string
Warning (emacs): 
ssh-manager.el:614: Argument `method' should appear (as METHOD) in the doc string
Warning (emacs): 
ssh-manager.el:197: Error messages should *not* end with a period
Warning (emacs): 
ssh-manager.el:280: Error messages should *not* end with a period
Warning (emacs): 
ssh-manager.el:288: Error messages should *not* end with a period
Warning (emacs): 
ssh-manager.el:347: Messages should start with a capital letter
Warning (emacs): 
ssh-manager.el:347: Error messages should *not* end with a period
Warning (emacs): 
ssh-manager.el:402: Error messages should *not* end with a period
Warning (emacs): 
ssh-manager.el:406: Error messages should *not* end with a period
Warning (emacs): 
ssh-manager.el:493: Error messages should *not* end with a period
Warning (emacs): 
ssh-manager.el:517: Messages should start with a capital letter
Warning (emacs): 
ssh-manager.el:603: Messages should start with a capital letter
```

Also:
- I don't fully understand your package, but are you aware of the builtin function `ibuffer-do-eval` and whether it captures some of what you're doing here?
- Running `(setq ssh-manager-mode-map (make-sparse-keymap))` at the top level will potentially clobber user definitions when they load your package.  Try: `(defvar my-map (let ((km (make-sparse-keymap))) (define-key ...) km))`
- The function `ssh-manager--init-term-mode` looks like a hacky implementation of a minor mode, which you should likely be using instead (with its own keymap, etc.)
- package-lint seems to have missed this, but `ssh-managerterm-kill-buffer-hook` should have a hyphen between `manager` and `term`
- `progn` is not needed nearly in as many places as you use it (e.g. surrounding entire function bodies)
- ssh-manager.el#L53: No `format` required; `message` takes an f-string -- very minor

7ym0n commented:
Thanks. 

I changed its.

![](https://github.com/7ym0n/dotfairy-ssh-manager/blob/master/ssh-manager.gif)

riscy commented:
Thanks for taking a look.  Based on the latest, I think you addressed a couple of the issues above, but not all of them.  Let me know if I'm mistaken.

7ym0n commented:
> Thanks for taking a look. Based on the latest, I think you addressed a couple of the issues above, but not all of them. Let me know if I'm mistaken.

Yes, checkdoc is not resolved

riscy commented:
Thank you -- please do your best to resolve them when you have a chance!

riscy commented:
I'll mark this awaiting upstream until you're ready to go.  But feel free to ping.

7ym0n commented:
> I'll mark this awaiting upstream until you're ready to go. But feel free to ping.

new latest version checkdoc is ok. I fixed.

riscy commented:
Thanks for checking those.  It's been a while, but I am still having trouble easily seeing what this package is all about.  I can see you compare it to `xshell` and `mobaxterm` but that doesn't tell me much.  If I trace through the code I can see numerous behaviors that feel like they shouldn't be part of the same package, and/or should be reconsidered:

- `ssh-manager-install-tools` which installs an executable called`sshpass` which seems like an important dependency of your package.  First, I'd suggest letting the users figure out how to install this tool themselves, especially since running a command like `install-tools` doesn't really tell them what's going on.  But more importantly, as I read more about sshpass, it strikes me as an ill-advised workaround for just setting up public key authentication on the server; so I would argue that this package is potentially instilling bad habits in terms of passing credentials around.  (If this were explicitly a wrapper for sshpass, like the package was named `sshpass-manager` for example, that would be different)
- `ssh-manager-keystroke-setup` which is about modifying `term-mode` keybindings.  I think this isn't so bad, but this package is after all an "ssh manager", not a "term manager", so I wonder if these are necessarily a good fit
- Commands like `ssh-manager-execute-current-line-cmd-to-ssh` and `ssh-manager-send-cmd-to-session-groups` which sends the same commands to several buffers in a group; but this doesn't seem like something that should apply only to ssh sessions, but rather a group of term-mode buffers?

And since I was just linting your file, a couple small suggestions pop out that weren't there before:
- ssh-manager.el#L266: Customizable variables shouldn't be private
- ssh-manager.el#L272: Customizable variables shouldn't be private

7ym0n commented:
This package is mainly used to create a session configuration, like xshell and other tools. The proxy server is required according to SSH, or TOTP secondary authentication is required for server connection. sshpass tool, which solves automatic password input and TOTP, can facilitate server management.

`ssh-manager-send-cmd-to-session-groups` is used to execute commands to a group of servers. It is often used for software installation, service deployment, etc. because it depends on SSH session,
I don't know whether other packages have been implemented, so I implemented a set of term-mode buffers in this package

Thanks!!! I Fixed.
	•	ssh-manager.el#L266: Customizable variables shouldn't be private
	•	ssh-manager.el#L272: Customizable variables shouldn't be private

riscy commented:
I feel not everything I wrote about sshpass was addressed, but admit I overloaded those bullet points.  So I apologize if some of my responses below seem a little repetitive.  

> sshpass tool, which solves automatic password input and TOTP, can facilitate server management.

To add to what I wrote about the installation of the sshpass dependency, also consider its [manpage](https://linux.die.net/man/1/sshpass) warns you that: "All system users can see the password in the command line with a simple "ps" command."  You may personally be aware of the risks, but a user could **easily** use this package without being aware of these implications.  I think you are even silently dumping unobscured cedentials to disk when you "persist" these sessions (without the user being aware of this).  I suggest you instead read up on Emacs' builtin auth-source for managing credentials (see functions like `auth-source-pick-first-password` and `auth-source-search`).  I don't think this feature is ready for MELPA.

> I don't know whether other packages have been implemented, so I implemented a set of term-mode buffers in this package.

I think the use of term-mode is not a blocker, although in that case I suggest calling this package ssh-term-manager to be a bit more specific about what the user is getting into.
